File: cetera.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-3.0-only or GPL-3.0-or-later
#
# Brief description of the software written in C or C++ or
#  or C# of Rust or Node.JS or a million other languages.
#
# Copyright (C) 2023 Andrew D. Harris <andrew2325@gmail.com>.
# Gemini was used to create this frontend to ect.  I am not a software engineer!
# This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.

# code...

import os
import subprocess
import tkinter as tk
from tkinter import filedialog, messagebox, ttk

class ECTGui(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("Cetera")
        self.geometry("560x520")
        self.configure(bg="#1e1e1e")

        style = ttk.Style()
        style.theme_use("clam")

        title_label = tk.Label(self, text="Cetera - ECT & Archive Tool", fg="#ffffff", bg="#1e1e1e", font=("Helvetica", 13, "bold"))
        title_label.pack(pady=12)

        # Main dynamic container frame that swaps views
        self.container_frame = tk.Frame(self, bg="#1e1e1e")
        self.container_frame.pack(pady=5, fill="x", padx=20)

        # Build View 1: ECT Controls & Mode Switcher
        self.build_ect_view()

        # Build View 2: Archive Selection Menu
        self.build_archive_menu_view()

        # Start on ECT View
        self.show_ect_view()

        # Log Output view (shared at the bottom)
        log_frame = tk.Frame(self, bg="#1e1e1e")
        log_frame.pack(pady=10, fill="both", expand=True, padx=20)

        self.log_text = tk.Text(log_frame, bg="#2d2d2d", fg="#ffffff", insertbackground="white", relief="flat", font=("Courier", 9))
        self.log_text.pack(side="left", fill="both", expand=True)

        scrollbar = tk.Scrollbar(log_frame, command=self.log_text.yview)
        scrollbar.pack(side="right", fill="y")
        self.log_text.config(yscrollcommand=scrollbar.set)

    def build_ect_view(self):
        self.ect_view_frame = tk.Frame(self.container_frame, bg="#1e1e1e")

        # Options Frame (ECT Level & Strip)
        opt_frame = tk.Frame(self.ect_view_frame, bg="#1e1e1e")
        opt_frame.pack(pady=5, fill="x")

        tk.Label(opt_frame, text="ECT Level (1-9):", fg="#cccccc", bg="#1e1e1e").pack(side="left", padx=5)
        self.level_spin = tk.Spinbox(opt_frame, from_=1, to=9, width=4)
        self.level_spin.delete(0, "end")
        self.level_spin.insert(0, "9")
        self.level_spin.pack(side="left", padx=5)

        self.strip_var = tk.BooleanVar(value=True)
        strip_chk = tk.Checkbutton(opt_frame, text="Strip Metadata", variable=self.strip_var, fg="#ffffff", bg="#1e1e1e", selectcolor="#333333", activebackground="#1e1e1e", activeforeground="#ffffff")
        strip_chk.pack(side="right", padx=5)

        # Action Buttons Frame
        btn_frame = tk.Frame(self.ect_view_frame, bg="#1e1e1e")
        btn_frame.pack(pady=10)

        file_btn = tk.Button(btn_frame, text="Select Files (ECT)", command=self.process_files, bg="#007acc", fg="white", padx=8, pady=6, relief="flat", font=("Helvetica", 9, "bold"))
        file_btn.pack(side="left", padx=4)

        dir_btn = tk.Button(btn_frame, text="Select Directory (ECT)", command=self.process_directory, bg="#28a745", fg="white", padx=8, pady=6, relief="flat", font=("Helvetica", 9, "bold"))
        dir_btn.pack(side="left", padx=4)

        switch_btn = tk.Button(btn_frame, text="Archive Menu...", command=self.show_archive_menu_view, bg="#d97706", fg="white", padx=8, pady=6, relief="flat", font=("Helvetica", 9, "bold"))
        switch_btn.pack(side="left", padx=4)

    def build_archive_menu_view(self):
        self.archive_menu_frame = tk.Frame(self.container_frame, bg="#1e1e1e")

        tk.Label(self.archive_menu_frame, text="Archive Creator", fg="#ffffff", bg="#1e1e1e", font=("Helvetica", 11, "bold")).pack(pady=5)

        # Format Selection Row
        fmt_frame = tk.Frame(self.archive_menu_frame, bg="#1e1e1e")
        fmt_frame.pack(pady=5)

        tk.Label(fmt_frame, text="Format:", fg="#cccccc", bg="#1e1e1e").pack(side="left", padx=5)
        self.archive_format_var = tk.StringVar(value="7z")
        tk.Radiobutton(fmt_frame, text="p7zip (.7z)", variable=self.archive_format_var, value="7z", fg="#ffffff", bg="#1e1e1e", selectcolor="#333333", activebackground="#1e1e1e", activeforeground="#ffffff").pack(side="left", padx=5)
        tk.Radiobutton(fmt_frame, text="tar.xz (.tar.xz)", variable=self.archive_format_var, value="tar.xz", fg="#ffffff", bg="#1e1e1e", selectcolor="#333333", activebackground="#1e1e1e", activeforeground="#ffffff").pack(side="left", padx=5)

        # Archive Action Buttons
        arc_btn_frame = tk.Frame(self.archive_menu_frame, bg="#1e1e1e")
        arc_btn_frame.pack(pady=10)

        arc_files_btn = tk.Button(arc_btn_frame, text="Select Files to Archive", command=self.archive_selected_files, bg="#007acc", fg="white", padx=10, pady=6, relief="flat", font=("Helvetica", 9, "bold"))
        arc_files_btn.pack(side="left", padx=6)

        arc_dir_btn = tk.Button(arc_btn_frame, text="Select Directory to Archive", command=self.archive_selected_directory, bg="#28a745", fg="white", padx=10, pady=6, relief="flat", font=("Helvetica", 9, "bold"))
        arc_dir_btn.pack(side="left", padx=6)

        back_btn = tk.Button(arc_btn_frame, text="< Back", command=self.show_ect_view, bg="#555555", fg="white", padx=10, pady=6, relief="flat", font=("Helvetica", 9, "bold"))
        back_btn.pack(side="left", padx=6)

    def show_ect_view(self):
        self.archive_menu_frame.pack_forget()
        self.ect_view_frame.pack(fill="x", expand=True)

    def show_archive_menu_view(self):
        self.ect_view_frame.pack_forget()
        self.archive_menu_frame.pack(fill="x", expand=True)

    def log(self, message):
        self.log_text.insert("end", message + "\n")
        self.log_text.see("end")

    def process_files(self):
        files = filedialog.askopenfilenames(
            title="Select Files to Optimize",
            filetypes=[("Supported Files", "*.png *.jpg *.jpeg *.zip *.gz"), ("All Files", "*.*")]
        )
        if not files:
            return
        self.run_ect_compression(list(files), "selected files")

    def process_directory(self):
        directory = filedialog.askdirectory(title="Select Directory to Optimize")
        if not directory:
            return

        supported_exts = (".png", ".jpg", ".jpeg", ".zip", ".gz")
        files_to_process = []
        
        for root, _, filenames in os.walk(directory):
            for filename in filenames:
                if filename.lower().endswith(supported_exts):
                    files_to_process.append(os.path.join(root, filename))

        if not files_to_process:
            messagebox.showinfo("No Files", "No supported files found in the selected directory.")
            return

        self.run_ect_compression(files_to_process, directory)

    def run_ect_compression(self, files_to_process, source_desc):
        if subprocess.run(["which", "ect"], capture_output=True).returncode != 0:
            messagebox.showerror("Error", "'ect' command not found.\nInstall it via source or your package manager.")
            return

        level = f"-{self.level_spin.get()}"
        strip = "-strip" if self.strip_var.get() else None

        self.log(f"Starting ECT compression for {len(files_to_process)} items from {source_desc} at level {level}...")
        self.update()

        success_count = 0
        fail_count = 0

        for file in files_to_process:
            cmd = ["ect", level]
            if strip:
                cmd.append(strip)
            cmd.append(file)

            while True:
                self.log(f"Executing: {' '.join(cmd)}")
                self.update()

                res = subprocess.run(cmd, capture_output=True, text=True)
                if res.returncode == 0:
                    self.log(f"Success: {os.path.basename(file)}")
                    success_count += 1
                    break
                else:
                    err = res.stderr.strip() or res.stdout.strip()
                    self.log(f"Error on {os.path.basename(file)}: {err}")
                    
                    retry = messagebox.askretrycancel(
                        "Compression Error", 
                        f"Failed to compress:\n{os.path.basename(file)}\n\nError: {err}\n\nClick 'Retry' to try again, or 'Cancel' to skip this file."
                    )
                    if retry:
                        continue
                    else:
                        self.log(f"Skipped: {os.path.basename(file)}")
                        fail_count += 1
                        break

        self.log(f"\nFinished ECT! Success: {success_count}, Failed/Skipped: {fail_count}")
        messagebox.showinfo("Done", f"ECT compression complete.\nSuccess: {success_count}\nFailed/Skipped: {fail_count}")

    def archive_selected_files(self):
        files = filedialog.askopenfilenames(title="Select Files to Archive")
        if not files:
            return
        
        fmt = self.archive_format_var.get()
        ext = ".7z" if fmt == "7z" else ".tar.xz"
        
        out_file = filedialog.asksaveasfilename(
            title="Save Archive As",
            initialfile=f"archive{ext}",
            defaultextension=ext,
            filetypes=[(f"{fmt.upper()} Archive", f"*{ext}"), ("All Files", "*.*")]
        )
        if not out_file:
            return

        self.run_archiver_files(list(files), out_file, fmt)

    def archive_selected_directory(self):
        src_dir = filedialog.askdirectory(title="Select Directory to Archive")
        if not src_dir:
            return

        fmt = self.archive_format_var.get()
        ext = ".7z" if fmt == "7z" else ".tar.xz"
        default_name = os.path.basename(os.path.normpath(src_dir)) + ext
        
        out_file = filedialog.asksaveasfilename(
            title="Save Archive As",
            initialfile=default_name,
            defaultextension=ext,
            filetypes=[(f"{fmt.upper()} Archive", f"*{ext}"), ("All Files", "*.*")]
        )
        if not out_file:
            return

        self.run_archiver_directory(src_dir, out_file, fmt)

    def run_archiver_files(self, files, out_file, fmt):
        if fmt == "7z":
            if subprocess.run(["which", "7z"], capture_output=True).returncode != 0:
                messagebox.showerror("Error", "'7z' command not found.\nInstall it via: sudo apt install p7zip-full")
                return
            cmd = ["7z", "a", out_file] + files
        else: # tar.xz using 9e compression level
            if subprocess.run(["which", "tar"], capture_output=True).returncode != 0:
                messagebox.showerror("Error", "'tar' command not found.")
                return
            os.environ["XZ_OPT"] = "-9e"
            cmd = ["tar", "-cJf", out_file] + files

        self.log(f"Starting archive creation: {' '.join(cmd)}")
        self.update()

        res = subprocess.run(cmd, capture_output=True, text=True)
        if "XZ_OPT" in os.environ:
            del os.environ["XZ_OPT"]

        if res.returncode == 0:
            self.log(f"Successfully created archive: {out_file}")
            messagebox.showinfo("Success", f"Archive created successfully:\n{out_file}")
        else:
            err = res.stderr.strip() or res.stdout.strip()
            self.log(f"Archive error: {err}")
            messagebox.showerror("Error", f"Failed to create archive:\n{err}")

    def run_archiver_directory(self, src_dir, out_file, fmt):
        if fmt == "7z":
            if subprocess.run(["which", "7z"], capture_output=True).returncode != 0:
                messagebox.showerror("Error", "'7z' command not found.\nInstall it via: sudo apt install p7zip-full")
                return
            cmd = ["7z", "a", out_file, src_dir]
        else: # tar.xz using 9e compression level
            if subprocess.run(["which", "tar"], capture_output=True).returncode != 0:
                messagebox.showerror("Error", "'tar' command not found.")
                return
            os.environ["XZ_OPT"] = "-9e"
            parent_dir = os.path.dirname(os.path.normpath(src_dir))
            dir_name = os.path.basename(os.path.normpath(src_dir))
            cmd = ["tar", "-cJf", out_file, "-C", parent_dir, dir_name]

        self.log(f"Starting archive creation: {' '.join(cmd)}")
        self.update()

        res = subprocess.run(cmd, capture_output=True, text=True)
        if "XZ_OPT" in os.environ:
            del os.environ["XZ_OPT"]

        if res.returncode == 0:
            self.log(f"Successfully created archive: {out_file}")
            messagebox.showinfo("Success", f"Archive created successfully:\n{out_file}")
        else:
            err = res.stderr.strip() or res.stdout.strip()
            self.log(f"Archive error: {err}")
            messagebox.showerror("Error", f"Failed to create archive:\n{err}")

if __name__ == "__main__":
    app = ECTGui()
    app.mainloop()

File: cetera.cpp
/*
* Copyright (C) 2026 [Your Name or Organization]
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <https://gnu.org>.
* Andrew D. Harris <andrew2325@gmail.com> used Gemini to make this app.  
* I am not an engineer or anything.
*/


#include <tcl.h>
#include <tk.h>
#include <cstdlib>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>
#include <sys/stat.h>
#include <dirent.h>

std::vector<std::string> ParseTclList(Tcl_Interp *interp, const char *listStr) {
    int argc = 0;
    const char **argv = nullptr;
    std::vector<std::string> result;
    if (Tcl_SplitList(interp, listStr, &argc, &argv) == TCL_OK) {
        for (int i = 0; i < argc; ++i) {
            result.push_back(argv[i]);
        }
        Tcl_Free((char *)argv);
    }
    return result;
}

void LogMessage(Tcl_Interp *interp, const std::string &msg) {
    std::string cmd = ".log_frame.txt insert end \"" + msg + "\\n\"\n";
    Tcl_Eval(interp, cmd.c_str());
    Tcl_Eval(interp, ".log_frame.txt see end\nupdate");
}

void CollectFiles(const std::string &dir, std::vector<std::string> &files) {
    DIR *dp = opendir(dir.c_str());
    if (!dp) return;
    struct dirent *entry;
    while ((entry = readdir(dp)) != nullptr) {
        std::string name = entry->d_name;
        if (name == "." || name == "..") continue;
        std::string fullpath = dir + "/" + name;
        struct stat st;
        if (stat(fullpath.c_str(), &st) == 0) {
            if (S_ISDIR(st.st_mode)) {
                CollectFiles(fullpath, files);
            } else if (S_ISREG(st.st_mode)) {
                size_t dot_pos = name.find_last_of('.');
                if (dot_pos != std::string::npos) {
                    std::string ext = name.substr(dot_pos);
                    for (auto &c : ext) c = tolower(c);
                    
                    if (ext == ".png" || ext == ".jpg" || ext == ".zip" || ext == ".gz" || ext == ".jpeg") {
                        files.push_back(fullpath);
                    }
                }
            }
        }
    }
    closedir(dp);
}

int ProcessECTCmd(ClientData clientData, Tcl_Interp *interp, int argc, const char *argv[]) {
    if (system("which ect > /dev/null 2>&1") != 0) {
        Tcl_Eval(interp, "tk_messageBox -icon error -title \"Error\" -message \"'ect' command not found.\\nInstall it via source or package manager.\"");
        return TCL_OK;
    }

    const char *mode = argv[1];
    std::vector<std::string> files_to_process;

    if (std::string(mode) == "files") {
        Tcl_Eval(interp, "tk_getOpenFile -multiple 1 -title \"Select Files to Optimize\" -filetypes {{\"Supported Files\" {*.png *.jpg *.jpeg *.zip *.gz}} {\"All Files\" {*.*}}}");
        std::string resStr = Tcl_GetString(Tcl_GetObjResult(interp));
        if (resStr.empty()) return TCL_OK;
        files_to_process = ParseTclList(interp, resStr.c_str());
    } else {
        Tcl_Eval(interp, "tk_chooseDirectory -title \"Select Directory to Optimize\"");
        std::string dirStr = Tcl_GetString(Tcl_GetObjResult(interp));
        if (dirStr.empty()) return TCL_OK;
        CollectFiles(dirStr, files_to_process);
        if (files_to_process.empty()) {
            Tcl_Eval(interp, "tk_messageBox -icon info -title \"No Files\" -message \"No supported files found in selected directory.\"");
            return TCL_OK;
        }
    }

    Tcl_Eval(interp, "set _lvl [.container.ect.opt.spn get]");
    std::string level = "-" + std::string(Tcl_GetString(Tcl_GetObjResult(interp)));
    
    Tcl_Eval(interp, "set _strip $strip_var");
    bool strip = (std::string(Tcl_GetString(Tcl_GetObjResult(interp))) == "1");

    LogMessage(interp, "Starting ECT compression for " + std::to_string(files_to_process.size()) + " items at level " + level + "...");

    int success_count = 0;
    int fail_count = 0;

    for (const auto &file : files_to_process) {
        std::string cmd = "ect " + level;
        if (strip) cmd += " -strip";
        cmd += " \"" + file + "\"";

        while (true) {
            LogMessage(interp, "Executing: " + cmd);
            int ret = system(cmd.c_str());
            
            size_t idx = file.find_last_of("/\\");
            std::string basename = (idx == std::string::npos) ? file : file.substr(idx + 1);

            if (ret == 0) {
                LogMessage(interp, "Success: " + basename);
                success_count++;
                break;
            } else {
                LogMessage(interp, "Error on " + basename);
                std::string box_cmd = "tk_messageBox -icon error -type retrycancel -title \"Compression Error\" -message \"Failed to compress:\\n" + basename + "\\n\\nClick 'Retry' to try again, or 'Cancel' to skip.\"";
                Tcl_Eval(interp, box_cmd.c_str());
                std::string choice = Tcl_GetString(Tcl_GetObjResult(interp));
                
                if (choice == "retry") {
                    continue;
                } else {
                    LogMessage(interp, "Skipped: " + basename);
                    fail_count++;
                    break;
                }
            }
        }
    }

    LogMessage(interp, "\nFinished ECT! Success: " + std::to_string(success_count) + ", Failed/Skipped: " + std::to_string(fail_count));
    Tcl_Eval(interp, ("tk_messageBox -icon info -title \"Done\" -message \"ECT compression complete.\\nSuccess: " + std::to_string(success_count) + "\\nFailed/Skipped: " + std::to_string(fail_count) + "\"").c_str());

    return TCL_OK;
}

int ProcessArchiveCmd(ClientData clientData, Tcl_Interp *interp, int argc, const char *argv[]) {
    const char *mode = argv[1];
    
    Tcl_Eval(interp, "set _fmt $archive_format_var");
    std::string fmt = Tcl_GetString(Tcl_GetObjResult(interp));
    std::string ext = (fmt == "7z") ? ".7z" : ".tar.xz";

    std::vector<std::string> targets;
    std::string default_name = "archive" + ext;

    if (std::string(mode) == "files") {
        Tcl_Eval(interp, "tk_getOpenFile -multiple 1 -title \"Select Files to Archive\"");
        std::string resStr = Tcl_GetString(Tcl_GetObjResult(interp));
        if (resStr.empty()) return TCL_OK;
        targets = ParseTclList(interp, resStr.c_str());
    } else {
        Tcl_Eval(interp, "tk_chooseDirectory -title \"Select Directory to Archive\"");
        std::string dirStr = Tcl_GetString(Tcl_GetObjResult(interp));
        if (dirStr.empty()) return TCL_OK;
        targets.push_back(dirStr);
        
        size_t idx = dirStr.find_last_of("/\\");
        std::string dirname = (idx == std::string::npos) ? dirStr : dirStr.substr(idx + 1);
        default_name = dirname + ext;
    }

    std::string dialog_cmd = "tk_getSaveFile -title \"Save Archive As\" -initialfile \"" + default_name + "\" -defaultextension " + ext + " -filetypes {{\"Archive\" {*" + ext + "}} {\"All Files\" {*.*}}}";
    if (Tcl_Eval(interp, dialog_cmd.c_str()) == TCL_ERROR) return TCL_OK;
    
    std::string out_file = Tcl_GetString(Tcl_GetObjResult(interp));
    if (out_file.empty()) return TCL_OK;

    std::string cmd = "";
    if (fmt == "7z") {
        if (system("which 7z > /dev/null 2>&1") != 0) {
            Tcl_Eval(interp, "tk_messageBox -icon error -title \"Error\" -message \"'7z' command not found.\\nInstall it via: sudo apt install p7zip-full\"");
            return TCL_OK;
        }
        cmd = "7z a \"" + out_file + "\"";
        for (const auto &t : targets) cmd += " \"" + t + "\"";
    } else {
        if (system("which tar > /dev/null 2>&1") != 0) {
            Tcl_Eval(interp, "tk_messageBox -icon error -title \"Error\" -message \"'tar' command not found.\"");
            return TCL_OK;
        }
        if (std::string(mode) == "dir") {
            std::string src = targets[0];
            size_t idx = src.find_last_of("/\\");
            std::string parent = (idx == std::string::npos) ? "." : src.substr(0, idx);
            std::string name = (idx == std::string::npos) ? src : src.substr(idx + 1);
            cmd = "XZ_OPT=-9e tar -cJf \"" + out_file + "\" -C \"" + parent + "\" \"" + name + "\"";
        } else {
            cmd = "XZ_OPT=-9e tar -cJf \"" + out_file + "\"";
            for (const auto &t : targets) cmd += " \"" + t + "\"";
        }
    }

    LogMessage(interp, "Starting archive creation: " + cmd);
    int ret = system(cmd.c_str());

    if (ret == 0) {
        LogMessage(interp, "Successfully created archive: " + out_file);
        Tcl_Eval(interp, ("tk_messageBox -icon info -title \"Success\" -message \"Archive created successfully:\\n" + out_file + "\"").c_str());
    } else {
        LogMessage(interp, "Archive error encountered.");
        Tcl_Eval(interp, "tk_messageBox -icon error -title \"Error\" -message \"Failed to create archive.\"");
    }

    return TCL_OK;
}

int main(int argc, char **argv) {
    Tcl_FindExecutable(argv[0]);
    Tcl_Interp *interp = Tcl_CreateInterp();

    if (Tcl_Init(interp) == TCL_ERROR) return 1;
    if (Tk_Init(interp) == TCL_ERROR) return 1;

    Tcl_CreateCommand(interp, "CppProcessECT", ProcessECTCmd, (ClientData)NULL, (Tcl_CmdDeleteProc *)NULL);
    Tcl_CreateCommand(interp, "CppProcessArchive", ProcessArchiveCmd, (ClientData)NULL, (Tcl_CmdDeleteProc *)NULL);

    std::string ui_script = 
        "wm title . \"cetera\"\n"
        "wm geometry . \"560x520\"\n"
        "wm resizable . 0 0\n"
        "option add *background #1e1e1e\n"
        "option add *foreground #ffffff\n"
        "option add *Label*background #1e1e1e\n"
        "option add *Label*foreground #ffffff\n"
        "option add *Labelframe*background #1e1e1e\n"
        "option add *Labelframe*foreground #cccccc\n"
        "option add *Radiobutton*background #1e1e1e\n"
        "option add *Radiobutton*foreground #ffffff\n"
        "option add *Radiobutton*selectColor #333333\n"
        "option add *Frame*background #1e1e1e\n"
        ". configure -background #1e1e1e\n"
        "label .title -text \"cetera - ECT & Archive Tool\" -font {Helvetica 13 bold}\n"
        "pack .title -pady 12\n"
        "frame .container\n"
        "pack .container -pady 5 -fill x -padx 20\n"
        "frame .container.ect\n"
        "frame .container.ect.opt\n"
        "pack .container.ect.opt -pady 5 -fill x\n"
        "label .container.ect.opt.lbl -text \"ECT Level (1-9):\" -fg #cccccc\n"
        "spinbox .container.ect.opt.spn -from 1 -to 9 -width 4\n"
        ".container.ect.opt.spn delete 0 end\n"
        ".container.ect.opt.spn insert 0 9\n"
        "pack .container.ect.opt.lbl -side left -padx 5\n"
        "pack .container.ect.opt.spn -side left -padx 5\n"
        "global strip_var\n"
        "set strip_var 1\n"
        "checkbutton .container.ect.opt.chk -text \"Strip Metadata\" -variable strip_var -selectcolor #333333 -activebackground #1e1e1e -activeforeground #ffffff\n"
        "pack .container.ect.opt.chk -side right -padx 5\n"
        "frame .container.ect.btns\n"
        "pack .container.ect.btns -pady 10\n"
        "button .container.ect.btns.file -text \"Select Files (ECT)\" -command {CppProcessECT files} -bg #007acc -fg white -padx 8 -pady 6 -relief flat -font {Helvetica 9 bold}\n"
        "button .container.ect.btns.dir -text \"Select Directory (ECT)\" -command {CppProcessECT dir} -bg #28a745 -fg white -padx 8 -pady 6 -relief flat -font {Helvetica 9 bold}\n"
        "button .container.ect.btns.switch -text \"Archive Menu...\" -command ShowArchiveView -bg #d97706 -fg white -padx 8 -pady 6 -relief flat -font {Helvetica 9 bold}\n"
        "pack .container.ect.btns.file -side left -padx 4\n"
        "pack .container.ect.btns.dir -side left -padx 4\n"
        "pack .container.ect.btns.switch -side left -padx 4\n"
        "frame .container.arc\n"
        "label .container.arc.lbl -text \"Archive Creator\" -font {Helvetica 11 bold}\n"
        "pack .container.arc.lbl -pady 5\n"
        "frame .container.arc.fmt\n"
        "pack .container.arc.fmt -pady 5\n"
        "label .container.arc.fmt.lbl -text \"Format:\" -fg #cccccc\n"
        "global archive_format_var\n"
        "set archive_format_var \"7z\"\n"
        "radiobutton .container.arc.fmt.r1 -text \"p7zip (.7z)\" -variable archive_format_var -value \"7z\" -selectcolor #333333 -activebackground #1e1e1e -activeforeground #ffffff\n"
        "radiobutton .container.arc.fmt.r2 -text \"tar.xz (.tar.xz)\" -variable archive_format_var -value \"tar.xz\" -selectcolor #333333 -activebackground #1e1e1e -activeforeground #ffffff\n"
        "pack .container.arc.fmt.lbl -side left -padx 5\n"
        "pack .container.arc.fmt.r1 -side left -padx 5\n"
        "pack .container.arc.fmt.r2 -side left -padx 5\n"
        "frame .container.arc.btns\n"
        "pack .container.arc.btns -pady 10\n"
        "button .container.arc.btns.farc -text \"Select Files to Archive\" -command {CppProcessArchive files} -bg #007acc -fg white -padx 10 -pady 6 -relief flat -font {Helvetica 9 bold}\n"
        "button .container.arc.btns.darc -text \"Select Directory to Archive\" -command {CppProcessArchive dir} -bg #28a745 -fg white -padx 10 -pady 6 -relief flat -font {Helvetica 9 bold}\n"
        "button .container.arc.btns.back -text \"< Back\" -command ShowECTView -bg #555555 -fg white -padx 10 -pady 6 -relief flat -font {Helvetica 9 bold}\n"
        "pack .container.arc.btns.farc -side left -padx 6\n"
        "pack .container.arc.btns.darc -side left -padx 6\n"
        "pack .container.arc.btns.back -side left -padx 6\n"
        "frame .log_frame\n"
        "pack .log_frame -pady 10 -fill both -expand 1 -padx 20\n"
        "text .log_frame.txt -bg #2d2d2d -fg #ffffff -insertbackground white -relief flat -font {Courier 9} -yscrollcommand {.log_frame.scr set}\n"
        "scrollbar .log_frame.scr -command {.log_frame.txt yview}\n"
        "pack .log_frame.txt -side left -fill both -expand 1\n"
        "pack .log_frame.scr -side right -fill y\n"
        "proc ShowECTView {} {\n"
        "    pack forget .container.arc\n"
        "    pack .container.ect -fill x -expand 1\n"
        "}\n"
        "proc ShowArchiveView {} {\n"
        "    pack forget .container.ect\n"
        "    pack .container.arc -fill x -expand 1\n"
        "}\n"
        "ShowECTView\n";

    if (Tcl_Eval(interp, ui_script.c_str()) == TCL_ERROR) {
        const char *err = Tcl_GetStringResult(interp);
        fprintf(stderr, "Tcl Error: %s\n", err);
        return 1;
    }

    Tk_MainLoop();
    return 0;
}
